sn_cli/src/subcommands/folders.rs

```diff
@@ -0,0 +1,315 @@
+            chunk_manager.chunk_path(&path, true, make_public)?;
+
+            let mut folders = build_folders_hierarchy(&path, client, root_dir)?;
+
+            // add chunked files to the corresponding Folders
+            for chunked_file in chunk_manager.iter_chunked_files() {
+                if let (Some(file_name), Some(parent)) = (
+                    chunked_file.file_name.to_str(),
+                    chunked_file.file_path.parent(),
+                ) {
+                    if let Some(folder) = folders.get_mut(parent) {
+                        folder.add_file(file_name.to_string(), chunked_file.head_chunk_address)?;
+                    }
+                }
+            }
+
+            println!("Paying for folders hierarchy and uploading...");
+            let root_dir_address = folders
+                .get(&path)
+                .map(|folder| *folder.address())
+                .ok_or(eyre!("Failed to obtain main Folder network address"))?;
+
+            pay_and_upload_folders(folders, verify_store, client, root_dir).await?;
+
+            println!(
+                "\nFolder hierarchy from {path:?} uploaded susccessfully at {}",
+                root_dir_address.to_hex()
+            );
+        }
+        FoldersCmds::Download {
+            folder_addr,
+            folder_name,
+            batch_size,
+        } => {
+            let address =
+                RegisterAddress::from_hex(&folder_addr).expect("Failed to parse Folder address");
+
+            let download_dir = dirs_next::download_dir().unwrap_or(root_dir.to_path_buf());
+            let download_folder_path = download_dir.join(folder_name);
+            println!(
+                "Downloading onto {download_folder_path:?} from {} with batch-size {batch_size}",
+                address.to_hex()
+            );
+            debug!(
+                "Downloading onto {download_folder_path:?} from {}",
+                address.to_hex()
+            );
+
+            let mut files_to_download = vec![];
+            let mut folders_to_download =
+                vec![("".to_string(), address, download_folder_path.clone())];
+
+            while let Some((name, folder_addr, target_path)) = folders_to_download.pop() {
+                if !name.is_empty() {
+                    println!(
+                        "Downloading Folder '{name}' from {}",
+                        hex::encode(folder_addr.xorname())
+                    );
+                }
+                download_folder(
+                    root_dir,
+                    client,
+                    &target_path,
+                    folder_addr,
+                    &mut files_to_download,
+                    &mut folders_to_download,
+                )
+                .await?;
+            }
+
+            let files_api: FilesApi = FilesApi::new(client.clone(), download_folder_path);
+            let uploaded_files_path = root_dir.join(UPLOADED_FILES);
```

Comment:
looks like it will try to download all `uploaded files` always ?
I thought it shall only try to download the files that within the `folder hierarchy` ?

Reply:
only those which belong to the folders being downloaded